### ISSUE TITLE
Add small message "ping/pong" benchmark & fix performance early in development

### ISSUE THREAD

**Issue by alexhultman** [denoland/deno#977]
Deno 0.1.8 is the first version I was able to benchmark. What I'm looking at is small message ping/pong between server/client over TCP with nodelay. This is where Node.js falls behind significantly compared to similar solutions (Golang, etc) and an important case to fix up in deno.

A small message is sent from client to server and from server to client, repeatedly as fast as possible. Server is deno, client is an optimized C++ client but could be deno itself. Currently only 1 connection is made, should be a few more when deno can handle them.

This checks the theoretical max performance of (HTTP) REST services sending small responses (a few kilobytes), websocket messaging (a few bytes) and similar small-IO cases.

I benchmarked C++/C++, Node.js v8.10.0/C++, Old Node.js v0.8.6/C++ and deno v0.1.8/C++.

The very old Node.js v0.8.6 does about 45k echoes/second, the new Node.js v8.10.0 does about 39k, deno v0.1.8 does 13k. Deno was however using 1.6 CPU cores 100%, while Node.js was only using 1 CPU core 100%.

This means currently deno cannot theoretically do more than 13k HTTP req/sec or 13k websocket echoes/second, which is just about 28% the efficiency of Node.js v0.8.6 (from 2012). If you normalize per spent CPU-time the percentage is about 17% of Node.js.

My point is that, performance is often times pretty much set in stone in the very first versions, and cannot be changed any significant amount if not fixed at an early stage (due to design flaws not fixed early enough). If things are not fixed in the early stage and tracked over time, deno will not succeed in being more efficient than node.js (which already is way behind Golang). Maybe this is not the goal, but I don't see why that shouldn't be a target.

Therefore I would like to see this IO case tracked and optimized for, together with the large message send benchmarking. Both cases should perform appropriately.

Asking Node.js devs to fix their performance problems today is met with "oh, you should have been here when we started, it's too late to make any of those changes now".

**Comment by hayd** [comment]
Would [wrk](https://github.com/wg/wrk) suffice to run the benchmark for this?

**Comment by alexhultman** [comment]
Yes but simplest possible and most self contained test would be to just ping/pong via localhost from deno client to deno server in the same script. Gradually increase message size from a few bytes to a few megabytes and plot a graph. Then do the same with a third party "target" such as Golang or Node.js depending on what goal you want to achieve.

**Comment by hayd** [comment]
Can you share your code / make a PR?

I tried to write something in deno but it's performance was all over the place across runs. (I think it's also using a ton of file descriptors and I was occasionally getting errors.)
So if you already have code for this it would be great!

wrk benchmark seemingly needs http (I can't seem to record results for TCP).

**Comment by ry** [comment]
Glad we’re finally able to measure response time. Thanks for looking into this. I agree that we need a benchmark like this running continuously on https://denoland.github.io/deno/
Sucks that Deno’s slower than Node at the moment - but I don’t think we’re locked into this perf at all - there’s a ton of low hanging fruit to pick on the optimization tree. 

I think using Deno as the client to benchmark the server isn’t such a good idea. I’m not confident that it can drive itself to capacity. It would be better to use wrk.

**Comment by alexhultman** [comment]
The following code "works" with wrk but isn't guaranteed to be correct always (it works well in practice though):

```js
import * as deno from "deno";
let listener = deno.listen("tcp", "127.0.0.1:3000");
const response = new TextEncoder().encode('HTTP/1.1 200 OK\r\nContent-Length: 12\r\n\r\nHello World!');

function acceptNew() {
	listener.accept().then(async (conn) => {

	acceptNew();

	const buffer = new Uint8Array(1024);

	try {
		while (true) {
			let result = await conn.read(buffer);
			if (result.eof) {
				break;
			}
			await conn.write(response);
			if (result.eof) {
				break;
			}
		}

	} catch (e) {

	}

	conn.close();
	});
}

acceptNew();
```

Here are the Deno results on my laptop:
```
alexhultman@alexhultman-N56VZ:~/wrk$ ./wrk http://localhost:3000
Running 10s test @ http://localhost:3000
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   546.59us  740.81us  16.01ms   96.41%
    Req/Sec    11.23k   758.26    12.82k    83.00%
  223507 requests in 10.00s, 10.87MB read
Requests/sec:  22348.85
Transfer/sec:      1.09MB
```

Here is Node.js:

```
alexhultman@alexhultman-N56VZ:~/wrk$ ./wrk http://localhost:3000
Running 10s test @ http://localhost:3000
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   249.74us  165.11us   9.56ms   99.00%
    Req/Sec    20.39k     2.14k   40.99k    93.53%
  407713 requests in 10.10s, 19.83MB read
Requests/sec:  40370.18
Transfer/sec:      1.96MB
```

Here is a reference result (third-party) on the same laptop, showing the margin for improvement in both Deno and Node.js:
```
alexhultman@alexhultman-N56VZ:~/wrk$ ./wrk http://localhost:3000
Running 10s test @ http://localhost:3000
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    43.93us   16.39us 433.00us   63.25%
    Req/Sec   109.96k     3.48k  119.03k    69.80%
  2209388 requests in 10.10s, 189.63MB read
Requests/sec: 218751.20
Transfer/sec:     18.78MB
```

I don't know how to take it from here so please take the code, fix it up and deploy to that graph.

**Comment by rsp** [comment]
Some time ago I made a simple but configurable benchmark for networking in Node.

It is configurable by environment variables of how many connections, how many writes per connection and how many bytes per write it does, and whether they are made serially or in parallel.

By default it connects to itself over the loopback interface but it can use different interfaces by setting IP address in the env vars.

It outputs a single value of milliseconds, I wrote it to be run by shell scripts with different settings and on different Node versions to compare the results.

I just put it on:
* https://github.com/rsp/node-net-benchmark
Maybe it would be useful to someone here.
